docs/Codebase.md

```diff
@@ -0,0 +1,96 @@
+    Enumerable.Empty<Regex>(),
+    ImmutableHashSet.Create("VS.ExternalAPIs.MSBuild"),
+    null,
+    null,
+    null);
+```
+
+#### [InsertionApi Class](#insertionapi-class)
+`InsertionApi` is the only implementation of `IInsertionApi` in the solution. The main logic executes the following steps in order:
+1. Load _default.config_ file into memory. All the operations on _default.config_ are handled in `DefaultConfigUpdater` class.
+1. Load manifest.json files.
+1. Determine the updates to the _default.config_ file. This operation is executed in parallel for each asset defined in manifest files.
+1. Determine the updates to `.props` files.
+    1. Determine all the `.swr` files.
+    2. Search the contents of `.swr` files for variables of the form `$(variable_name)`.
+    3. Download the packages updated in step 3 into memory.
+    4. Determine the value of the variables by comparing the _file names in the nuget packages_ with _the file names listed in swr payloads_.
+1. In the declaring `.props` files, replace the value of the variables with the values found in the previous step.
+1. Flush changes to `.props` files.
+1. Flush changes to `default.config` and `.packageconfig` files.
+
+### [Console Project](#console-project)
+The console project contains the command line user interface codebase and logging.
+
+#### [Program Class](#program-class)
+- The Program class contains the assembly entry point
+and is the most interesting class in the assembly
+- The class defines the switches, explained in the _[Input](docs/README.md#input)_ section.
+
+#### [Logging](#logging)
+The application's logging solution relies on _System.Diagnostics.Trace_ events & registers _ConsoleTraceListener_ 
+and `TextWriterTraceListener` listeners, respectively.  See the _[Output](docs/README.md#output)_ section for more details.
+
+### [Build Pipeline](#build-pipeline)
+The build definition resides in the [`azure-pipelines.yml`](/azure-pipelines.yml) file.
+There are two build pipelines for InsertionsClient:
+- [Public pipeline](https://dev.azure.com/dnceng/public/_build?definitionId=846&_a=summary). Runs the validation builds for PRs.
+
+- [Internal pipeline](https://dev.azure.com/dnceng/internal/_build?definitionId=847&_a=summary). Besides validation, generates and publishes nuget packages to the [feed](https://dev.azure.com/dnceng/public/_packaging?_a=package&feed=dotnet-eng). There are two packages published:
+  - [Microsoft.DotNet.InsertionsClient.Core](https://dev.azure.com/dnceng/public/_packaging?_a=package&feed=dotnet-eng&package=Microsoft.DotNet.InsertionsClient.Core&protocolType=NuGet)
+  - [Microsoft.DotNet.InsertionsClient.Console](https://dev.azure.com/dnceng/public/_packaging?_a=package&feed=dotnet-eng&package=Microsoft.DotNet.InsertionsClient.Console&protocolType=NuGet)
+
+  Major, minor and patch version numbers are specified in the `azure-pipelines.yml` file. A unique build number is calculated for each run and appended to the version number.
```

Comment:
Is paragraph needed here?

Reply:
This document is mainly intended for people maintaining this tool. In case they want to update version numbers, I wanted to mention how they can find them.